tik/lib/CartographerKernel.cpp

```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
-            int64_t id = GetBlockID(block);
-            if (blocks.find(id) == blocks.end() && block != Exit && block != Init && block != Exception && id != -2)
+            if (key.first->getName()[0] == 'i')
             {
-                for (auto user : block->users())
-                {
-                    if (auto *phi = dyn_cast<PHINode>(user))
-                    {
-                        phi->removeIncomingValue(block);
-                    }
-                    else
-                    {
-                        user->replaceUsesOfWith(block, Exit);
-                    }
-                }
-                bToRemove.push_back(block);
+                VMap[IDToValue[key.second]] = key.first;
             }
         }
-        /*
-    for (auto block : bToRemove)
-    {
-        //this breaks hard for some reason
-        //not really necessary fortunately
-        //block->eraseFromParent();
-    }
-    */
     }
 
-    void CartographerKernel::RemapNestedKernels(llvm::ValueToValueMapTy &VMap)
+    void CartographerKernel::BuildKernelFromBlocks(llvm::ValueToValueMapTy &VMap, set<BasicBlock *> &blocks)
     {
-        // Now find all calls to the embedded kernel functions in the body, if any, and change their arguments to the new ones
-        std::map<Argument *, Value *> embeddedCallArgs;
-        for (auto &bf : *(KernelFunction))
+        set<Function *> headFunctions;
+        for (const auto &ent : Entrances)
+        {
+            BasicBlock *eTarget = IDToBlock[ent->Block];
+            headFunctions.insert(eTarget->getParent());
+        }
+
+        if (headFunctions.size() != 1)
+        {
+            throw AtlasException("Entrances not on same level");
+        }
+
+        for (auto block : blocks)
         {
-            for (BasicBlock::iterator i = bf.begin(), BE = bf.end(); i != BE; ++i)
+            if (headFunctions.find(block->getParent()) == headFunctions.end())
             {
-                if (auto *callInst = dyn_cast<CallInst>(i))
+                continue;
+            }
+            int64_t id = GetBlockID(block);
+            if (KernelMap.find(id) != KernelMap.end())
+            {
+                //this belongs to a subkernel
+                auto nestedKernel = KernelMap[id];
+                bool inNested = false;
+                for (const auto &ent : nestedKernel->Entrances)
+                {
+                    if (IDToBlock[ent->Block] == block)
+                    {
+                        inNested = true;
+                        break;
+                    }
+                }
+                if (inNested)
                 {
-                    auto calledFunc = callInst->getCalledFunction();
-                    auto subK = KfMap[calledFunc];
-                    if (subK != nullptr)
+                    //we need to make a unique block for each entrance (there is currently only one)
+                    for (uint64_t i = 0; i < nestedKernel->Entrances.size(); i++)
```

Comment:
The ideal answer is to create a select statement that the phi will select if either entrance is used.

As long as this work we can leave it though. I think the optimizer should fix it ultimately anyway.

Reply:
You're saying we should create 1 block that enters the child from all entrances of a child, instead of a unique block for all entrances of a child

Reply:
Ideally if there are multiple "Entrances" that effectively enter the same block we don't add additional blocks. This isn't for all entrances, just for entrances that share a target. As I said, this can be skipped, but may make the code harder to understand in both c++ and LLVM.

Reply:
considering this resolved